core/server/api/v2/utils/serializers/output/utils/members.js

```diff
@@ -1,15 +1,39 @@
+const common = require('../../../../../../lib/common');
 const labs = require('../../../../../../services/labs');
+const membersService = require('../../../../../../services/members');
 const MEMBER_TAG = '#members';
+const PERMIT_CONTENT = false;
+const BLOCK_CONTENT = true;
 
 // Checks if request should hide memnbers only content
 function hideMembersOnlyContent(attrs, frame) {
-    let hasMemberTag = false;
-    if (labs.isSet('members') && !frame.original.context.member && attrs.tags) {
-        hasMemberTag = attrs.tags.find((tag) => {
-            return (tag.name === MEMBER_TAG);
-        });
+    const membersEnabled = labs.isSet('members');
+    if (!membersEnabled) {
+        return PERMIT_CONTENT;
+    }
+
+    const postHasMemberTag = attrs.tags && attrs.tags.find((tag) => {
+        return (tag.name === MEMBER_TAG);
+    });
+    const requestFromMember = frame.original.context.member;
+    if (!postHasMemberTag) {
+        return PERMIT_CONTENT;
+    }
+    if (!requestFromMember) {
+        return BLOCK_CONTENT;
+    }
+
+    const planRequired = membersService.api.paymentConfigured;
+    const memberHasPlan = !!(frame.original.context.member.plans || []).length;
+    if (!planRequired) {
+        return PERMIT_CONTENT;
+    }
+    if (memberHasPlan) {
+        return PERMIT_CONTENT;
     }
-    return hasMemberTag;
+    throw new common.errors.NoPermissionError({
```

Comment:
A shame to error in the serialization layer, but there is no output validation layer

Reply:
Why can't this be handled in the permission layer (e.g. permissible fn)? 🤔

Reply:
Because permissions happen before the query is run, and we need access to the tags to determine this state.

When members permissions are handled by the permissions module, we should be able to move this there :relaxed:

Reply:
🤔 But the permissible fn can ensure getting access to the tags of the target post?

Reply:
Hmm, so permissible would make a db query? Does that mean we'd make two queries for every request?

Reply:
We already do it e.g. look at the permissible fn's.